lib/rubygems/commands/rebuild_command.rb

```diff
@@ -0,0 +1,262 @@
+      Dir.chdir(build_path) { build_gem(gem_name, source_date_epoch, new_file) }
+    else
+      build_gem(gem_name, source_date_epoch, new_file)
+    end
+
+    compare(source_date_epoch, old_file, new_file)
+  end
+
+  private
+
+  def sha256(file)
+    Digest::SHA256.hexdigest(File.read(file))
+  end
+
+  def get_timestamp(file)
+    mtime = nil
+    Gem::Package::TarReader.new(File.open(file)) do |tar|
+      mtime = tar.seek("metadata.gz") {|f| f.header.mtime }
+    end
+
+    mtime
+  end
+
+  def compare(source_date_epoch, old_file, new_file)
+    date = Time.at(source_date_epoch.to_i).strftime("%F %T %Z")
+
+    old_hash = sha256(old_file)
+    new_hash = sha256(new_file)
+
+    say
+    say "Built at: #{date} (#{source_date_epoch})"
+    say "Original build saved to:   #{old_file}"
+    say "Reproduced build saved to: #{new_file}"
+    say "Working directory: #{options[:build_path] || Dir.pwd}"
+    say
+    say "Hash comparison:"
+    say "  #{old_hash}\t#{old_file}"
+    say "  #{new_hash}\t#{new_file}"
+    say
+
+    if old_hash == new_hash
+      say "SUCCESS - original and rebuild hashes matched"
+    else
+      say "FAILURE - original and rebuild hashes did not match"
+      terminate_interaction 1
+    end
+  end
+
+  def prep_dirs
+    rebuild_dir = File.expand_path("rebuild")
+    old_dir = File.join(rebuild_dir, "old")
+    new_dir = File.join(rebuild_dir, "new")
+
+    if File.directory?(rebuild_dir)
+      FileUtils.remove_dir(rebuild_dir)
+    end
+
+    FileUtils.mkdir_p(old_dir)
+    FileUtils.mkdir_p(new_dir)
+
+    [old_dir, new_dir]
+  end
+
+  def get_gem_name_and_version
+    args = options[:args] || []
+    if args.length == 2
+      gem_name, gem_version = args
+    elsif args.length > 2
+      raise Gem::CommandLineError, "Too many arguments"
+    else
+      raise Gem::CommandLineError, "Expected GEM_NAME and GEM_VERSION arguments (gem rebuild GEM_NAME GEM_VERSION)"
+    end
+
+    [gem_name, gem_version]
+  end
+
+  def build_gem(gem_name, source_date_epoch, output_file)
+    gemspec = options[:gemspec_file] || find_gemspec("#{gem_name}.gemspec")
+
+    if gemspec
+      build_package(gemspec, source_date_epoch, output_file)
+    else
+      alert_error error_message(gem_name)
+      terminate_interaction(1)
+    end
+  end
+
+  def build_package(gemspec, source_date_epoch, output_file)
+    with_source_date_epoch(source_date_epoch) do
+      spec = Gem::Specification.load(gemspec)
+      if spec
+        Gem::Package.build(
+          spec,
+          options[:force],
+          options[:strict],
+          output_file
+        )
+      else
+        alert_error "Error loading gemspec. Aborting."
+        terminate_interaction 1
+      end
+    end
+  end
+
+  def with_source_date_epoch(source_date_epoch)
+    old_sde = ENV["SOURCE_DATE_EPOCH"]
+    ENV["SOURCE_DATE_EPOCH"] = source_date_epoch.to_s
+
+    yield
+  ensure
+    ENV["SOURCE_DATE_EPOCH"] = old_sde
+  end
+
+  def find_gemspec(glob = "*.gemspec")
```

Comment:
can this share an implementation with the `build` command?

Reply:
Yeah, it looks like it's identical, so it can be shared.

Any idea where that should be put? It seems like shared things in `rubygems/commands/` just get dumped in `rubygems/*`, which seems like a pattern that may be worth worth breaking. Perhaps creating `rubygems/commands/shared/` for that purpose? (I'm okay with other names for such a directory, it's just first one that came to mind.)

<details><summary>details about shared things for `rubygems/commands/*`</summary>

For context, it appears that at least these files are only used in `rubygems/commands/` (and corresponding tests):

- `rubygems/gemcutter_utilities`
- `rubygems/install_message`
- `rubygems/install_default_message`
- `rubygems/install_update_options`
- `rubygems/installer`
- `rubygems/local_remote_options`
- `rubygems/query_utils`
- `rubygems/rdoc`

</details>